plugins/linux-capture/xcomposite-helpers.cpp

```diff
@@ -0,0 +1,569 @@
+	free(check2);
+
+	return true;
+}
+
+std::list<xcb_window_t> xcomp_top_level_windows(xcb_connection_t *conn)
+{
+	std::list<xcb_window_t> res;
+
+	// ewmh top level window listing is not supported.
+	if (ATOM__NET_CLIENT_LIST == XCB_ATOM_NONE)
+		return res;
+
+	xcb_screen_iterator_t screen_iter =
+		xcb_setup_roots_iterator(xcb_get_setup(conn));
+	for (; screen_iter.rem > 0; xcb_screen_next(&screen_iter)) {
+		xcb_generic_error_t *err = nullptr;
+		// Read properties up to 4096*4 bytes
+		xcb_get_property_cookie_t cl_list_cookie =
+			xcb_get_property(conn, 0, screen_iter.data->root,
+					 ATOM__NET_CLIENT_LIST, 0, 0, 4096);
+		xcb_get_property_reply_t *cl_list =
+			xcb_get_property_reply(conn, cl_list_cookie, &err);
+		DEFER(free(cl_list));
+		if (err != nullptr)
+			continue;
+
+		uint32_t len = xcb_get_property_value_length(cl_list) /
+			       sizeof(xcb_window_t);
+		for (uint32_t i = 0; i < len; i++)
+			res.push_back(((xcb_window_t *)xcb_get_property_value(
+				cl_list))[i]);
+	}
+
+	return res;
+}
+
+xcb_window_t xcomp_find_window(xcb_connection_t *conn, Display *disp,
+			       const char *str)
+{
+	std::list<xcb_window_t> tlw = xcomp_top_level_windows(conn);
+	if (!str || strlen(str) == 0) {
+		return tlw.front();
+	}
+	std::string wstr = str;
+
+	size_t firstMark = wstr.find(WIN_STRING_DIV);
+	size_t lastMark = wstr.rfind(WIN_STRING_DIV);
+	size_t markSize = strlen(WIN_STRING_DIV);
+
+	// wstr only consists of the window-id
+	if (firstMark == std::string::npos)
+		return (xcb_window_t)std::stol(wstr);
+
+	// wstr also contains window-name and window-class
+	xcb_window_t wid = std::stol(wstr.substr(0, firstMark));
+	std::string wname = wstr.substr(firstMark + markSize,
+					lastMark - firstMark - markSize);
+	std::string wcls = wstr.substr(lastMark + markSize);
+
+	// first try to find a match by the window-id
+	for (xcb_window_t cwin : tlw) {
+		if (cwin == wid) {
+			return cwin;
+		}
+	}
+
+	// then try to find a match by name & class
+	for (xcb_window_t cwin : tlw) {
+		if (wname == xcomp_window_name(conn, disp, cwin) &&
+		    wcls == xcomp_window_class(conn, cwin)) {
+			return cwin;
+		}
+	}
+
+	blog(LOG_DEBUG,
+	     "Did not find new window id for Name '%s' or Class '%s'",
+	     wname.c_str(), wcls.c_str());
+	return wid;
+}
+
+void xcomp_cleanup_pixmap(Display *disp, struct xcompcap *s)
+{
+	if (s->gltex) {
+		GLuint gltex = *(GLuint *)gs_texture_get_obj(s->gltex);
+		glBindTexture(GL_TEXTURE_2D, gltex);
+		if (s->glxpixmap) {
+			if (s->strict_binding) {
+				glXReleaseTexImageEXT(disp, s->glxpixmap,
+						      GLX_FRONT_EXT);
+			}
+			glXDestroyPixmap(disp, s->glxpixmap);
+			s->glxpixmap = 0;
+		}
+		gs_texture_destroy(s->gltex);
+		s->gltex = 0;
+	}
+
+	if (s->pixmap) {
+		XFreePixmap(disp, s->pixmap);
+		s->pixmap = 0;
+	}
+
+	if (s->tex) {
+		gs_texture_destroy(s->tex);
+		s->tex = 0;
+	}
+}
+
+static gs_color_format gs_format_from_tex()
+{
+	GLint iformat = 0;
+	// consider GL_ARB_internalformat_query
```

Comment:
Is this a to-do?

Reply:
This was a prexisting comment. That query will let us know the real format to avoid our hacks but it requires a very recent version of opengl so its unlikely to be reasonable.